pyop2/compilation.py

```diff
@@ -498,13 +501,34 @@ def __init__(self, code, argtypes):
         raise CompilationError("Don't know what compiler to use for platform '%s'" %
                                platform)
     dll = compiler.get_so(code, extension)
+    if code.local_kernel.events:
+        dll = _add_profiling_events(dll, code.local_kernel.events)
 
     fn = getattr(dll, fn_name)
     fn.argtypes = code.argtypes
     fn.restype = restype
     return fn
 
 
+def _add_profiling_events(dll, events):
+    """
+        If PyOP2 is in profiling mode, events are attached to dll to profile the local linear algebra calls.
+        The event is generated here in python and then set in the shared library,
+        so that memory is not allocated over and over again in the C kernel. The naming
+        convention is that the event ids are named by the event name prefixed by "ID_".
+    """
+    if PETSc.Log.isActive():
+        # also link the events from the linear algebra callables
+        if hasattr(dll, "solve"):
+            events += ('solve_memcpy', 'solve_getrf', 'solve_getrs')
+        if hasattr(dll, "inverse"):
+            events += ('inv_memcpy', 'inv_getrf', 'inv_getri')
+        # link all ids in DLL to the events generated here in python
+        for e in events:
+            ctypes.c_int.in_dll(dll, 'ID_'+e).value = PETSc.Log.Event(e).id
+    return dll
```

Comment:
I would call this an anti-pattern (possibly I'm wrong). If you are modifying `dll` in-place it should not also be returned from the function.

Reply:
Yes will change it

Reply:
It doesn't look like you've applied the change.

Reply:
Argh, my bad, I must have dropped my change when I squashed.

Reply:
Okay changed now.